goto_commands.py

```diff
@@ -0,0 +1,122 @@
+import sublime
+import sublime_plugin
+
+from itertools import dropwhile, takewhile
+from .lint import persist
+
+
+"""
+Implement typical Goto Next Previous Error Commands. If you will, add
+the following to your key bindings.
+
+```
+  { "keys": ["ctrl+k", "n"],
+    "command": "sublime_linter_goto_error",
+    "args": {"direction": "next"}
+  },
+  { "keys": ["ctrl+k", "p"],
+    "command": "sublime_linter_goto_error",
+    "args": {"direction": "previous"}
+  },
+```
+
+Supported args:
+
+* `direction`: 'previous' | 'next'
+* `count`: How wide you will jump. Defaults to 1.
+* `wrap`: Set to True, to jump to the top if you're on the last error and
+          vice versa. Defaults to False
+
+"""
+
+
+class SublimeLinterGotoError(sublime_plugin.WindowCommand):
+    def run(self, direction='next', count=1, wrap=False):
+        goto(self.window.active_view(), direction, count, wrap)
+
+
+def goto(view, direction, count, wrap):
+    bid = view.buffer_id()
+
+    try:
+        errors = persist.raw_errors[bid]
+    except KeyError:
+        flash(view, 'No errors/problems')
+        return
+
+    if len(errors) == 0:
+        flash(view, 'No errors/problems')
+        return
+
+    current_line, current_col = get_current_pos(view)
+    if (current_line, current_col) == (-1, -1):
+        return
+
+    # Filter out duplicates on the same position. Also filter out errors
+    # under the cursor, bc we don't want to jump to them.
+    errors = {(error['line'], error['start'])
+              for error in errors
+              if not(error['line'] == current_line and
+                     error['start'] <= current_col <= error['end'])}
+    errors = sorted(errors)
+
+    # Edge case: Since we filtered, it is possible we get here with nothing
+    # left. That is the case if we sit on the last remaining error, where we
+    # don't have anything to jump to and even `wrap` becomes a no-op.
+    if len(errors) == 0:
+        flash(view, 'No more problems.')
+        return
+
+    def before_current_pos(error):
+        line, start = error
+        return (
+            line < current_line or
+            # `start < current_col` is only safe bc we already
+            # filtered errors directly under cursor (!)
+            line == current_line and start < current_col)
+
+    next_positions = dropwhile(before_current_pos, errors)
+    previous_positions = takewhile(before_current_pos, errors)
+
```

Comment:
Here we iterate over errors twice. What about slicing `errors` based on `next_positions`?

Reply:
These are lazy. Both lines are declarative.